pyanaconda/dbus/interface.py

```diff
@@ -0,0 +1,323 @@
+
+    def __init__(self, xml_generator=XMLGenerator()):
+        self.xml_generator = xml_generator
+
+    def generate_specification(self, cls, interface_name=None):
+        """Generates DBus XML specification for given class.
+
+        If class defines a new interface, it will be added to
+        the specification.
+
+        :param cls: class object to decorate
+        :param str interface_name: name of the interface defined by class
+        :return str: DBus specification in XML
+        """
+        # Collect all interfaces that class inherits.
+        interfaces = self._collect_interfaces(cls)
+
+        # Generate a new interface.
+        if interface_name:
+            interface = self._generate_interface(cls, interfaces, interface_name)
+            interfaces[interface_name] = interface
+
+        # Generate XML specification for the given class.
+        node = self._generate_node(cls, interfaces)
+        return self.xml_generator.element_to_xml(node)
+
+    def _is_exportable(self, member_name):
+        """Is the name of a class member exportable?
+
+        The name is exportable if it follows the DBus specification.
+        Only CamelCase names are allowed.
+        """
+        return bool(self.NAME_PATTERN.fullmatch(member_name))
+
+    def _is_defined(self, interfaces, member_name):
+        """Is the member name defined in given interfaces?
+
+        :param interfaces: a dictionary of interfaces
+        :param member_name: a name of the class member
+        :return: True if the name is defined, otherwise False
+        """
+        return any(self.xml_generator.is_member(member)
+                   and self.xml_generator.has_name(member, member_name)
+                   for interface in interfaces.values()
+                   for member in interface)
+
+    def _iterate_parameters(self, member):
+        """Iterate over method parameters.
+
+        For every parameter returns its name, a type hint and a direction.
+
+        :param member: a method object
+        :return: an iterator
+        """
+        # Get type hints for parameters.
+        type_hints = get_type_hints(member)
+
+        # Get parameters.
+        parameters = inspect.signature(member).parameters
+
+        for index, name in enumerate(parameters):
+            # Skip self parameter.
+            if index < 1:
+                continue
+
+            # Check the kind of the parameter
+            if parameters[name].kind != Parameter.POSITIONAL_OR_KEYWORD:
+                raise DBusSpecificationError("Only positional or keyword arguments are allowed.")
+
+            # Check if the type is defined.
+            if name not in type_hints:
+                raise DBusSpecificationError("Parameter %s doesn't have defined type." % name)
+
+            yield name, type_hints[name], self.DIRECTION_IN
+
+        # Is the return type defined?
+        if self.RETURN_PARAMETER not in type_hints:
+            return
```

Comment:
That kind of check doesn't exist. Functions always return at least `None`. We can force programmers to always define a type hint for the return value and use `None` to indicate, it doesn't return anything.

For example:
```
def Function(self) -> None:
    pass
```